lib/Validator/Constraints/FieldDefinitionDefaultValueValidator.php

```diff
@@ -51,4 +52,39 @@ protected function generatePropertyPath($errorIndex, $errorTarget)
     {
         return 'defaultValue';
     }
+
+    /**
+     * Returns the field definition $value refers to.
+     * FieldDefinition object is needed to validate field value against field settings.
+     *
+     * @param FieldData|ValueObject $value ValueObject holding the field value to validate, e.g. FieldDefinitionData.
+     *
+     * @throws \InvalidArgumentException If field definition cannot be retrieved.
+     *
+     * @return \eZ\Publish\API\Repository\Values\ContentType\FieldDefinition
+     */
+    protected function getFieldDefinition(ValueObject $value)
+    {
+        return $this->getUpdatedFieldDefinition($value);
+    }
+
+    /**
+     * This method overwrite a property fieldSettings in the FieldDefinition object to expose sent settings.
+     *
+     * @param \eZ\Publish\API\Repository\Values\ValueObject $value
+     *
+     * @return \eZ\Publish\Core\Repository\Values\ContentType\FieldDefinition
+     */
+    private function getUpdatedFieldDefinition(ValueObject $value): FieldDefinition
+    {
+        $oldFieldDef = $value->fieldDefinition;
+
+        $properties = [];
+        foreach ($oldFieldDef->attributes() as $property) {
```

Comment:
This is well defined structure, so just explicitly use properties to initialize new modified value.
`attributes` is deprecated and at some point might disappear.

Note that depending on Core `FieldDefinition` is risky and unsupported, so you should at least add a unit test checking either this specific behavior or `FieldDefinition` structure itself. That way we can at least identify potential issue faster. (I'm more in favor of testing if this Validator returns proper updated structure).

Reply:
Ok I see you copied my solution from slack but I purposely used `getProperties()` instead of `attributes()` because later is deprecated as @alongosz noticed.